### ISSUE TITLE
Spike: Research design (UX/UI) options for find & replace

### ISSUE THREAD

**Issue by mlewand** [ckeditor/ckeditor5#9716]
We need to figure what options we have to implement find & replace and what pros or cons proposed options present.

Our reference regarding functionality is CKEditor 4's find & replace implementation - but this doesn't mean that we need to mimic it  1:1 when it comes to UI/UX implementation.

**Comment by magda-chrzescian** [comment]
Below are some thoughts about the possible options and behaviors. I also created a prioritization graph, which should be reviewed, especially the effort axis: https://miro.com/app/board/o9J_lDDL3Ok=/

### Must have

- Search input + search button
- Replace
    - Replace single occurrence
    - Replace all occurrences
- Match case option
- Match whole words only option
- Match cyclic option (from CKE4)?
    - Rarely used in other text editors, usually just always active. What's the case where we need to disable it?
    - Maybe not required if the amount of results is displayed so a user can easily tell if he is looking at the last occurrence?
- Find next
    - Usually executed with `Enter`
    - Do we want to have a separate button for that action (">") or execute it with the main "find" button?
- Find previous
    - Usually executed with the `Shift+Enter`
    - Do we want to have a separate button for that or execute it with the keyboard only?
- Highlighting search results and the currently selected result
- Fully keyboard-navigable UI
    - Potential conflicts with the navigation through the results?

### Nice to have

- Searching in multiple instances of the editor ([https://ckeditor.com/docs/ckeditor5/latest/examples/builds/inline-editor.html](https://ckeditor.com/docs/ckeditor5/latest/examples/builds/inline-editor.html))
- Searching in selection
    - The user has to set a selection in the balloon editor to display the toolbar, so searching in the selection has to be optional.
- Searching for a specific HTML/model elements (e.g. headings/bold text only)
- Ability to play with the text while the find&replace UI is open
    - Maybe the search results should remain "active" after closing the UI?
        - The "x" closes the drop-down and hides the results, focusing the editable closes the drop-down, but leaves the results (button marked as active).
        - Maybe the drop-down shouldn't close at all when focusing an editable? But it would be inconsistent with the other drop-down behavior and the drop-down can cover some search results.
    - Easy to achieve with the pop-up UI
    - Auto-refreshing search results when typing
- Emphasizing the search result which is currently being selected with an animation
- Displaying the amount of results and a currently highlighted result
- Replacement preview
- Making it easy to insert some special characters (tab, enter) - dedicated drop-down
- Allowing using RegExps (option)
    - and making it easy to insert them - dedicated drop-down
- Ignore punctuation/white spaces option
- Preserve case when replacing option
- Searching only in proposed changes option
- Ignore Latin diacritics (e. g. ä = a, E = É) option
- Keyboard shortcut
    - Where to display the UI if there is no button in the toolbar?
    - Do we want to override the browser search shortcut?
- Search only after/before the caret option
- Remembering the state (previous search conditions)
- Mini-outline to mark the search results
- Extendable UI - ability to add custom options

### Questions

- UI with tabs or a single sheet?
- Search on keyup or execute a search action with a button?
- A separate command for the find and the replacement, so the developer can use two buttons if needed?
- How should the UI be displayed?
    - Drop-down
        - 👍🏻 Easy to implement
        - 👍🏻 Doesn't take a lot of space in the UI (only a button)
        - ❓ Maybe the drop-down could be draggable?
        - 👎🏿 Potentially can cover a part of the document outline which can be crucial for interaction with the search results
        - 🔻 Can not be displayed if a respective button is not present in the UI (with a shortcut or command)
        - 🔻 Confusing if searching in multiple editor instances
    - Pop-up
        - 👍🏻 Doesn't take a lot of space in the UI (only a button)
        - 👍🏻 Easy to use with multiple editor instances
        - 👍🏻 Can be moved outside the editable area to not cover anything
        - 👍🏻 Easy to work with the results with constantly opened pop-up
        - 🔻 Hard to implement
    - Inline in the toolbar
        - 👍🏻 Extremely easy to use (for searching, not replacing)
        - 👍🏻 Replacing can be optional
        - 🔻 Hard to implement
        - 🔻 Takes a lot of space in the toolbar
        - 🔻 Would still require a drop-down to see custom settings and replacement options

**Comment by mlewand** [comment]
Thanks for solid research @magda-chrzescian that's a good conversation starter.

I would add few things:

*   multiroot support - we need to remember about this
*   mobile experience - this is something we should take into consideration when considering the UI approach

Regarding the keys, what about <kbd>f3</kbd> and <kbd>shift</kbd> + <kbd>f3</kbd>?  I know these are also commonly used to go to next match / go to previous match. I believe this is also something you've been thinking about as you brought the question "Do we want to override the browser search shortcut?".

**Comment by magda-chrzescian** [comment]
A TODO from the last meeting is to find some solution for a find&replace drop-down in the inline editor.
